src/SFA.DAS.Commitments.Application/Services/V2EventsPublisher.cs

```diff
@@ -24,117 +26,192 @@ public V2EventsPublisher(IEndpointInstance endpointInstance, ICommitmentsLogger
-                    ev.CreatedOn = _currentDateTime.Now;
-                    ev.Uln = apprenticeshipEvent.Apprenticeship.ULN;
-                    ev.ProviderId = apprenticeshipEvent.Apprenticeship.ProviderId;
-                    ev.AccountId = apprenticeshipEvent.Apprenticeship.EmployerAccountId;
-                    ev.AccountLegalEntityPublicHashedId =
-                        apprenticeshipEvent.Apprenticeship.AccountLegalEntityPublicHashedId;
-                    ev.LegalEntityName = apprenticeshipEvent.Commitment.LegalEntityName;
-                    ev.StartDate = apprenticeshipEvent.Apprenticeship.StartDate.Value;
-                    ev.EndDate = apprenticeshipEvent.Apprenticeship.EndDate.Value;
-                    ev.PriceEpisodes = priceEpisodes;
-                    ev.TrainingType = (CommitmentsV2.Types.TrainingType)apprenticeshipEvent.Apprenticeship.TrainingType;
-                    ev.TrainingCode = apprenticeshipEvent.Apprenticeship.TrainingCode;
-                    ev.TransferSenderId = apprenticeshipEvent.Apprenticeship.TransferSenderId;
-                });
+        private enum ApprenticePreChecks
+        {
+            NotRequired = 1,
+            HasStartDate = 2,
+            HasEndDate = 4,
+            HasStopDate = 8,
+            HasStartAndEndDate = HasStartDate | HasEndDate
+        }
+
+        private Task PublishWithLog<TEvent>(Action<TEvent> messageConstructor, string message) where TEvent : class
+        {
+            return PublishWithLog(ApprenticePreChecks.NotRequired, messageConstructor, message);
+        }
+
+        /// <summary>
+        ///     publish the specified message and log whether the publish was successful or not.
+        /// </summary>
+        /// <typeparam name="TEvent"></typeparam>
+        /// <param name="checks">Specified properties will be checked for null value before calling the message constructor</param>
+        /// <param name="messageConstructor">Sets values on an instance of the message</param>
+        /// <param name="message">A log message that will be recorded with the success or failure message</param>
+        /// <returns></returns>
+        private async Task PublishWithLog<TEvent>(ApprenticePreChecks checks, Action<TEvent> messageConstructor, string message) where TEvent : class
+        {
+            var logMessage = $"Publish {typeof(TEvent).Name} message. {message}";
+            try
+            {
+                await _endpointInstance.Publish<TEvent>(messageConstructor);
 
                 _logger.Info($"{logMessage} successful");
             }
             catch (Exception e)
             {
                 _logger.Error(e, $"{logMessage} failed");
+                throw;
             }
         }
 
-        public async Task PublishApprenticeshipStopped(Commitment commitment, Apprenticeship apprenticeship)
+        private void DoPreChecks<TEvent>(ApprenticePreChecks checks, Apprenticeship apprenticeship) where TEvent : class
```

Comment:
I like this DoPreChecks(), but it's not being used.  I still think we should keep it in, as I can see it being used in the very near future